tests/test_trtllm_mnnvl_allreduce.py

```diff
@@ -0,0 +1,227 @@
+
+    finally:
+        # Ensure cleanup happens even if assertions fail
+        del mcast_buffer_mnnvl
+
+
+"""Main test function that runs on each MPI rank"""
+
+
+@pytest.mark.parametrize("seq_len", [4])
+@pytest.mark.parametrize("fusion", [False])
+def test_mnnvl_allreduce_full(monkeypatch, seq_len: int, fusion: bool):
+    monkeypatch.setenv("TRTLLM_FORCE_MNNVL_AR", "1")  # force multi-node allreduce.
+
+    # Get MPI info
+    rank = MPI.COMM_WORLD.Get_rank()
+    world_size = MPI.COMM_WORLD.Get_size()
+
+    # Ensure we have exactly 2 ranks for this test
+    if world_size < 2:
+        if rank == 0:
+            print(f"ERROR: This test requires at least 2 MPI ranks, got {world_size}")
+        sys.exit(1)
+
+    # Set CUDA device based on rank
+    torch.cuda.set_device(rank)
+
+    if rank == 0:
+        print(f"Running MNNVL AllReduce test with {world_size} ranks")
+        print(f"Rank {rank} using GPU {torch.cuda.current_device()}")
+
+    hidden_size = 7168
+    dtype = torch.bfloat16
+    tensor_parallel_size = world_size
+    eps = 1e-5
+
+    torch.manual_seed(42)
+
+    try:
+        if rank == 0:
+            print(
+                f"Testing seq_len={seq_len}, hidden_size={hidden_size}, fusion={fusion}"
+            )
+
+        # Generate test data (same on all ranks due to same seed)
+        x_full = torch.randn((tensor_parallel_size, seq_len, hidden_size), dtype=dtype)
+        residual = torch.randn((seq_len, hidden_size), dtype=dtype)
+        norm_weight = torch.randn((hidden_size,), dtype=dtype)
+
+        # Each rank gets its slice of the input
+        x = x_full[rank, :, :]
+
+        # Compute reference output based on fusion mode
+        if fusion:
+            raise NotImplementedError("Fusion not implemented")
+        else:
+            # Non-fused case: Only AllReduce
+            allreduce_result = torch.sum(x_full, dim=0)  # AllReduce result
+            reference_output = (allreduce_result,)
+
+        # Run the test
+        row_linear_residual_norm_fusion_forward(
+            x,
+            residual,
+            norm_weight,
+            eps,
+            hidden_size,
+            dtype,
+            tensor_parallel_size,
+            rank,
+            fusion,
+            reference_output,
+        )
+
+        # Synchronize before next test
+        comm.mpi_barrier()
+
+        print(f"PASSED[rank={rank}]: seq_len={seq_len}, fusion={fusion}")
+
+    except Exception as e:
+        print(f"FAILED[rank={rank}]: seq_len={seq_len}, fusion={fusion} failed: {e}")
+        if rank == 0:
+            traceback.print_exc()
+        # Don't exit immediately, let other tests run
+        comm.mpi_barrier()
+
+    # Final synchronization and results
+    comm.mpi_barrier()
+
+
+if __name__ == "__main__":
+    # Test parameters
+    # seq_lens = [1, 4, 32, 128]
+    fusion_modes = [False]  # Only non-fused case for now
+    for fusion in fusion_modes:
+        test_mnnvl_allreduce_full(4, fusion)
```

Comment:
This will need to be updated because of the `monkeypatch` fixture (there should be 3 arguments now). Or it can be removed if the intention is to always run this w/ `pytest`